aips/aip-x.md

```diff
@@ -0,0 +1,192 @@
+
+### Out of scope
+
+We will only be covering the intent declaration and execution logic. We are not going to touch the solving logic on how intents should be fulfilled and we will leave it for builders on Aptos to figure out.
+
+The current design of the intent framework facilitates on chain declaration of intents to simplify the problem. Ideally the intent should be declarable offchain as well using signed message but this is beyond the scope of the current AIP. We will provide a path on how we can enable features like this in the future.
+
+The intent system designed here also focus on the Aptos only setup and we are not looking into cross chain solutions so far.
+
+## High-level Overview
+
+We see intent as a programmable lock for on chain assets. Meaning:
+- One party can declare:
+  - Type of Asset they would like to trade in
+  - Condition when this asset could be released
+- The framework provides mechanism to release the lock when such condition is met.
+
+Logic here can then be translated into a Move struct that looks like the following:
+```
+struct TradeIntent<Source, phantom Target, Args> has key {
+    offered_resource: Source,
+    argument: Args,
+    // Fn(Target, Args) -> ()
+    consumption_function: FunctionInfo,
+    self_delete_ref: DeleteRef,
+    expiry_time: u64,
+}
+```
+
+The `TradeIntent` is parameterized by three type parameters:
+1. `Source`: Value that intent issuer is willing to give away
+2. `Target`: the type of value that the intent issuer is expecting.
+3. `Argument`: Intent user will use `Argument` as auxillary info to determine whether the unlock condition is met.
+
+In the `TradeIntent`, we also register for a function pointer that can be set by the user, called `consumption_function`. The idea is that a transaction could only be committed if the intent issuer receives a value of `Target` type and also `consumption_function(target, argument)` is executed successfully. 
+
+The lifecycle of an intent will look like the following:
+1. User sends a transaction to the blockchain to create the trading intent by specifying the resource they can offer and the expected condition when such resource can be unlocked.
+2. The intent would be broadcasted using an event.
+3. Anyone can execute this intent as long as the `consumption_function` can be executed successfully.
+    - We will expect intent solvers to actively monitor the available intents on chain and create transactions to claim the intent.
+
+With those types being generic, we can implement different trading intents easily:
+- If `Target` is instantiated with `FungibleAsset`, it means we are trading a Move value for a specific fungible asset.
+- If `Target` is instantiated with `LinearTransferRef`, it means we are trading a Move value for an Object.
```

Comment:
Should we include `NonFungibleAsset` as a type as well? Certain DeFi protocols can mint an NFT for a certain staking/deposit operation, no?

Reply:
Here are just examples. You can put arbitrary type here.